### ISSUE TITLE
DPDK support

### ISSUE THREAD

**Issue by ghost** [netty/netty#13059]

**Comment by zzwxh** [comment]
f-stack or seastar

**Comment by hyperxpro** [comment]
PRs are welcome!

**Comment by normanmaurer** [comment]
Currently not planed at all.

**Comment by RainM** [comment]
@normanmaurer @hyperxpro I'm looking for enabling DPDK (f-stack) for netty. Is netty project is interested in such change? I try to implement such thing. So, I'd like to clarify if such contribution is welcome and would be merged. Such change will require 1) new version of transport like epoll's one 2) new event loop (based on Epoll's one).

**Comment by RainM** [comment]
For now, the most difficult thing is remove timer and wakeup fd from epoll's event because f-stack doesn't support all these things.

**Comment by hyperxpro** [comment]
> @normanmaurer @hyperxpro I'm looking for enabling DPDK (f-stack) for netty. Is netty project interested in such change? I try to implement such thing. So, I'd like to clarify if such contribution is welcome and would be merged. Such change will require 1) new version of native transport like epoll's one 2) new event loop java class (based on Epoll's one).

Definitely I'd be better to have separate transport based on Epoll with F-Stack. Because the performance difference will be massive, we'd love to have it.

**Comment by chrisvest** [comment]
@RainM if you intend to work on that, I have some recommendations:

First, I'd advise to do this in a separate incubator repository, similar to how the [io_uring](https://github.com/netty/netty-incubator-transport-io_uring) transport was implemented. This will give the code time to mature, and makes any sort of review work much easier since you can split it over as many PRs as it takes. This also gives you your own release schedule, independent of Netty, so it's easier to iterate and let people try it out.

Second, unless you specifically need this to work for Netty 4.1, I'd recommend integrating with 4.2. The low-level transport implementation APIs changed a bit, to better accommodate io_uring, so this is an area where we're _not_ preserving the compatibility between 4.1 and 4.2.

**Comment by RainM** [comment]
ok. Let us discuss internally if we start such development.

**Comment by RainM** [comment]
@normanmaurer 

I still dig deeper and deeper and get more issues (as expected, unfortunately). As far as I see, f-stack is just some 'library' on top of DPDK and doesn't abstract us from DPDK full way. From our perspective it means DPDK is fully responsible for thread lifecycle. And this lifecycle isn't correspond for netty's one. So, there are a lot of obstacles and unexpected thing.

Nevertheless, I got some measurements.
I've written simple echo server and test client with netty. Client connects to echo server, send current timestamp and waits for reply. Therefore, I measure round trip time from client to echo server and back. 
I've ported echo server for working with DPDK/F-Stack. Therefore, there are two versions of echo server. Pure netty (on top of kernel's epoll) and with my transport (on top of F-Stack's epoll)

Test setup:
1. First VM in AWS (c6i group) with tester application
2. Second VM with two network interfaces. The first one is managed by default kernel. Default echo server application uses this network interface. Second interface is managed by DPDK. My hacked version of echo server works with this interfac.

Result:

1. Default implementation (kernel) timings (in nanoseconds):
```
p50: 64831 // 64us    <<<<<< median
p90: 70847                 <<<<<< quantile 0.90
p99: 98175                <<<<<< quantile 0.99
p999: 110015
p9999: 465151
p99999: 2328575 // 2.3ms
```

2. Versino with dpdk-enabled echo server:
```
p50: 53087 // 53 us      <<<<<<<< median
p90: 56991                   <<<<<< quantile 0.90
p99: 66495                  <<<<<< quantile 0.99
p999: 71679
p9999: 106751
p99999: 144127 // 144us
```


So, As far as I see there is no times increase in median. But high quantile (rare cases) shows great improvement.

Does it sound interesting? There are a lot of thin places in possible places which we need to discuss how to iplement this better way.

BTW, I disabled ipv6, domain sockets and UDP for a while. First target is tcp+ipv4 only.

**Comment by hyperxpro** [comment]
@RainM 

That's right, F-Stack is supposed to provide a high-level abstraction to DPDK, making integration easy. Can you show the difference in throughput b/w DPDK and Epoll and Iouring? Also point me to repo if it's public.

Thanks a lot for bringing this up!

**Comment by RainM** [comment]
No, repo isn't public yet. Regarding throughput.. Our application is latency sensitive. It's why I measure latency. Neveertheless, How to measure throughput proper way? Just transfer as much bytes as I can via one connection? Run multiple connections and read/write many bytes via every connection?

**Comment by hyperxpro** [comment]
HTTP RPS will do it.

**Comment by chrisvest** [comment]
Reopening in fairness to the discussion and work.

**Comment by RainM** [comment]
Regarding implementation. There are some features of DPDK which give us strange things in netty.

- DPDK creates thread on it's own. As far as I see there is no way to provide something like ThreadFactory or poll DPDK/F-stack from our own threads. There is CPU mask parameter in fstack's config. Not sure how it works yet.
- F-Stack accepts text file-based config with addresses, routing, sysctls (https://github.com/F-Stack/f-stack/blob/dev/config.ini). 
- I still don't understand how f-stack handles multithreading. (there is lcore_mark argument in f-stack's config. don't know if causes spawning multiple thread). Nevertheless, I see issue in F-stack repo with discussion regarding multithreading (https://github.com/F-Stack/f-stack/issues/834). As far as I understand, F-Stack implies for users to read data from single thread. Someone has attached diffs for running multiple f-stack based network stacks in one process (one per thread? not sure).

Summarising this, what I'm doing and with which constraints.
1. Some DSL which generates F-stack's config. Java code should provide IP addresses, mask, gateway and so on. Quite simple DSL for the beginning.
2. F-stack initialization could be done during JNI library loading or some static call from java code. I'd prefer second option for the sake of simplicity and single responsibility. Config file should be ready by this time.
3. DPDK creates thread during initialization. I will rewrite f-stack bootstrup code (https://github.com/F-Stack/f-stack/blob/dev/lib/ff_dpdk_if.c#L2041). Modified version will register new thread for JVM via AttachCurrentThread JNI call.
4. Since DPDK create event loop thread on his own, I need to remember this thread somehow. Next, java's ThreadFactory for F-Stack-enabled event loop should return this single thread only. Therefore, the only supported event loop thread count will be one.

**Comment by franz1981** [comment]
Just curious: have your tried already openonload kernel bypass using Netty as-it-is already?

**Comment by RainM** [comment]
Open onload supports either solarflare network card or AF_XDP kernel bypass. There is no solarflare network card available in AWS EC2. AF_XDP itself isn't oriented for low latency. Details: https://github.com/Xilinx-CNS/onload/issues/139

**Comment by RainM** [comment]
@chrisvest @normanmaurer @hyperxpro what do you thing about my last notes regarding making ThreadFactory which will return only one thread? It's ugly, but I see no other option. netty + DPDK will be single-threaded (at least IO part).

**Comment by chrisvest** [comment]
Netty event loops are themselves single-threaded. Reusing the same thread across multiples of them will not be a good idea. I think it'll be better to enforce that the corresponding `MultithreadEventLoopGroup` only has one thread, by making the `ThreadFactory` throw if it's asked to create more than one thread.

I thought the threading model for DPDK was quite similar to io_uring though, except completions are delivered through callbacks instead of through a completion queue? As such I expect we could keep the Netty threads what they are, and have them interact with the DPDK system with some task hand-over. Have I misunderstood?

**Comment by RainM** [comment]
You are right with one note to mention. DPDK doesn't let external threads to interact with them. For example, there are a lot of thread-local variable in DPDK's threads. And it makes really difficult to read DPDK's data from some other thread. 

So, we can't just spawn thread via java's thread factory and let this thread read data from DPDK/F-stack. The only DPDK's threads are allowed to read data from DPDK.

So, as I write before I need to create DPDK's thread, remember it somehow. After that java's ThreadFactory should return this thread only.

**Comment by chrisvest** [comment]
@RainM There isn't, but you can get the fd with `intValue()` and implement your own calls. The io_uring integration does this.

**Comment by RainM** [comment]
May be it's better to remove final keyword from these methods? What is the purpose to keep them from overriding?
For example, EpollChannel are derived from UnixChannel class. And this interface just returns FileDescriptor instance with wrong methods. And anybody could call them directly. Not sure if it's proper behaviour. I'd prefer to override methods from FileDescriptor.

**Comment by RainM** [comment]
BTW, close method of FileDescriptor isn't final. And looks like it's ok. I see no reason to have methods in FileDescriptor to be final.

```
    public void close() throws IOException {
        if (markClosed()) {
            int res = close(fd);
            if (res < 0) {
                throw newIOException("close", res);
            }
        }
    }
```

**Comment by chrisvest** [comment]
If DPDK don't use regular file descriptors like the others, then yeah I guess you need to override those methods. Feel free to open a PR.

**Comment by RainM** [comment]
@chrisvest 
BTW, why do you use hawtjni maven plugin? Why not just directly execute some Makefile? 
I've faced issue with this plugin. Actually issue is somewhere in libtool called by this plugin.
I need to pass a bunch of linker arguments from Maven't pom file. Flags are obtained from pkg-config tool. Flags make ld linker to include full static archives into result binary (DSO in my case). In particular, these flags are something like '-Wl,--whole-archive ...[long list of static libraries] ... -Wl,--no-whole-archive'. And suddenly libtool sorts these flags somehow and all linker flags (-Wl,*) turn ont to be in the end of command line.
So, instead of arguments like 
-Wl,--whole-archive ... -ldpdk ... -Wl,--no-whole-archive -lnuma
result command args look like
 ... -ldpdk ... -lnuma -Wl,--whole-archive -Wl,--no-whole-archive  [look at order of arguments]

As a result, DPDK's driver static libraries aren't linked into result executable DSO. Moreover, for example libnuma's dependency are not resolved as well because of incorrect order of params.

Workaround I used is to transform a list of arguments into one long linker argument like:

-Wl,--whole-archive,...,-ldpdk,...,--no-whole-archive,-lnuma [spaces are replaced with commas, as a result libtool/gcc's driver recognizes this as one long argument to LD]

It works but looks really ridiculous.


And besides all this issues hawtjni doesn't handle source code modification and works REALLY LONG time (because each rebuild requires 'configure' phase). As a result one need to manually remove all build files and wait whole rebuild of DSO.

**Comment by chrisvest** [comment]
@RainM Netty was using hawtjni long before I joined the project, so I don't know.

**Comment by normanmaurer** [comment]
@RainM it's been a while... it worked for us so we sticked with it. It also made things easier with support windows etc. That said I am happy to revisit if someone makes a good proposal.

**Comment by normanmaurer** [comment]
Let me close this for now as there is really no plan to add official support atm
